system/jlib/jiface.hpp

```diff
@@ -119,31 +119,46 @@ class CInterfaceOf : public CSimpleInterfaceOf<INTERFACE>
 public:
     virtual void beforeDispose() {}
 
-    inline bool isAlive() const         { return this->xxcount.load(std::memory_order_relaxed) < DEAD_PSEUDO_COUNT; }       //only safe if Link() is called first
+    //Function to be called when checking if an entry in a non-linking cache is valid
+    //must be called inside a critical section that protects access to the cache.
+    //This function must only increment the link count if the object is valid - otherwise it
+    //end up being freed more than once.
+    inline bool isAliveAndLink() const
+    {
+        unsigned expected = this->xxcount.load(std::memory_order_acquire);
+        for (;;)
+        {
+            //If count is 0, or count >= DEAD_PSEUDO_COUNT then return false - combine both into a single test
+            if ((expected-1) >= (DEAD_PSEUDO_COUNT-1))
+                return false;
+
+            //Avoid incrementing the link count if expected=0, otherwise it introduces a race condition
```

Comment:
the comment confused me a little, as expected=0 already dealt with above, perhaps comment needs changing to just:
```
  // compare and exchange to increment to avoid race condition
```

.. or similar

Reply:
yes, should say if xxcount=0.  because the value may have changed to 0 in the meantime.
I will modify

Reply:
changed comment